neo/core/dataobject.py

```diff
@@ -0,0 +1,365 @@
+    elif not isinstance(value, (list, np.ndarray)) or \
+            (isinstance(value, pq.Quantity) and value.shape == ()):
+        _check_annotations(value)
+        value = _normalize_array_annotations(np.array([value]), length)
+
+    # If array annotation, check for correct length,
+    # only single dimension and allowed data
+    else:
+        # Get length that is required for array annotations, which is equal to the length
+        # of the object's data
+        own_length = length
+
+        # Escape check if empty array or list and just annotate an empty array (length 0)
+        # This enables the user to easily create dummy array annotations that will be filled
+        # with data later on
+        if len(value) == 0:
+            if not isinstance(value, np.ndarray):
+                value = np.ndarray((0,))
+            val_length = own_length
+        else:
+            # Note: len(o) also works for np.ndarray, it then uses the outmost dimension,
+            # which is exactly the desired behaviour here
+            val_length = len(value)
+
+        if not own_length == val_length:
+            raise ValueError("Incorrect length of array annotation: {} != {}".
+                             format(val_length, own_length))
+
+        # Local function used to check single elements of a list or an array
+        # They must not be lists or arrays and fit the usual annotation data types
+        def _check_single_elem(element):
+            # Nested array annotations not allowed currently
+            # So if an entry is a list or a np.ndarray, it's not allowed,
+            # except if it's a quantity of length 1
+            if isinstance(element, list) or \
+                    (isinstance(element, np.ndarray) and not
+                    (isinstance(element, pq.Quantity) and element.shape == ())):
+                raise ValueError("Array annotations should only be 1-dimensional")
+            if isinstance(element, dict):
+                raise ValueError("Dicts are not supported array annotations")
+
+            # Perform regular check for elements of array or list
+            _check_annotations(element)
+
+        # Arrays only need testing of single element to make sure the others are the same
+        if isinstance(value, np.ndarray):
+            # Type of first element is representative for all others
+            # Thus just performing a check on the first element is enough
+            # Even if it's a pq.Quantity, which can be scalar or array, this is still true
+            # Because a np.ndarray cannot contain scalars and sequences simultaneously
+            try:
+                # Perform check on first element
+                _check_single_elem(value[0])
+            except IndexError:
+                # If length of data is 0, then nothing needs to be checked
+                pass
+            return value
+
+        # In case of list, it needs to be ensured that all data are of the same type
```

Comment:
Doesn't it make more sense to first convert to an np.ndarray? Converting elements to a consistent dtype would then be already handled by this.